[UI] openshift-storage namespace not found on storage Clients in Provider mode

The issue in title is failing most of the UI tests
 
 @ui
    @runs_on_provider
    @tier1
    @polarion_id("OCS-4925")
    @skipif_ui_not_support("validation")
    def test_storage_cluster_validation_ui(self, setup_ui_class):
        """
        Validate Storage Cluster status on UI
    
        Args:
            setup_ui_class: login function on conftest file
    
        """
        validation_ui_obj = ValidationUI()
&gt;       validation_ui_obj.validate_storage_cluster_ui()

tests/cross_functional/ui/test_validation_ui.py:51: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
ocs_ci/ocs/ui/validation_ui.py:562: in validate_storage_cluster_ui
    self.select_namespace(project_name="openshift-storage")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ValidationUI Web Page, project_name = 'openshift-storage'

    def select_namespace(self, project_name):
        """
        This function selects the namespace on UI.
        The timeout is hard-coded to 10 seconds in the below function call which is more than sufficient.
    
        Args:
            project_name (str): Name of the project to be selected
    
        Returns:
            bool: True if the project is found, raises NoSuchElementException otherwise with a log message
        """
    
        from ocs_ci.ocs.ui.helpers_ui import format_locator
    
        self.do_click(self.generic_locators["project_selector"])
    
        # if project is already selected, skip and proceed further
        if self.get_elements(
            format_locator(self.generic_locators["project_selected"], project_name)
        ):
            self.take_screenshot("namespace_selected")
            logger.info("Project already selected")
    
            self.do_click(self.generic_locators["project_selector"])
            return True
    
        default_projects_is_checked = self.wait_for_element_to_be_present(
            self.generic_locators["show_default_projects_toggle"]
        )
        if default_projects_is_checked.get_attribute("data-checked-state") == "false":
            logger.info("Show default projects")
            self.do_click(self.page_nav["show-default-projects"])
    
        logger.info(f"Wait and select namespace {project_name}")
        wait_for_project = self.wait_until_expected_text_is_found(
            locator=format_locator(
                self.generic_locators["test-project-link"], project_name, project_name
            ),
            expected_text=f"{project_name}",
            timeout=10,
        )
        if wait_for_project:
            self.do_click(
                format_locator(
                    self.generic_locators["test-project-link"],
                    project_name,
                    project_name,
                )
            )
            logger.info(f"Namespace {project_name} selected")
            return True
        else:
&gt;           raise NoSuchElementException(f"Namespace {project_name} not found on UI")
E           selenium.common.exceptions.NoSuchElementException: Message: Namespace openshift-storage not found on UI

ocs_ci/ocs/ui/page_objects/page_navigator.py:400: NoSuchElementException</failure></testcase><testcase classname="tests.cross_functional.ui.test_validation_ui.TestUserInterfaceValidation" name="test_dashboard_validation_ui" time="77.202"><properties><property name="squad" value="Black" /><property name="polarion-testcase-id" value="OCS-2575" /></properties><failure message="selenium.common.exceptions.TimeoutException: Message: Failed to find the element (xpath,//span[text()='openshift-storage'])">_locator = ("//span[text()='openshift-storage']", 'xpath'), _timeout = 30
_enable_screenshot = True, _copy_dom = False

    def _do_click(_locator, _timeout=30, _enable_screenshot=False, _copy_dom=False):
        # wait for page fully loaded only if an element was not located
        # prevents needless waiting and frequent crushes on ODF Overview page,
        # when metrics and alerts frequently updated
        if not self.get_elements(_locator):
            self.page_has_loaded()
        screenshot = (
            ocsci_config.UI_SELENIUM.get("screenshot") and enable_screenshot
        )
        if screenshot:
            self.take_screenshot()
        if _copy_dom:
            self.copy_dom()
    
        wait = WebDriverWait(self.driver, timeout)
        try:
            if (
                version.get_semantic_version(get_ocp_version(), True)
                &lt;= version.VERSION_4_11
            ):
                element = wait.until(
                    ec.element_to_be_clickable((locator[1], locator[0]))
                )
            else:
&gt;               element = wait.until(
                    ec.visibility_of_element_located((locator[1], locator[0]))
                )

ocs_ci/ocs/ui/base_ui.py:202: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = &lt;selenium.webdriver.support.wait.WebDriverWait (session="03c971acaa084890f609985623f58b0a")&gt;
method = &lt;selenium.webdriver.support.expected_conditions.visibility_of_element_located object at 0x7f7f3afc5f70&gt;
message = ''

    def until(self, method, message=''):
        """Calls the method provided with the driver as an argument until the \
        return value is not False."""
        screen = None
        stacktrace = None
    
        end_time = time.time() + self._timeout
        while True:
            try:
                value = method(self._driver)
                if value:
                    return value
            except self._ignored_exceptions as exc:
                screen = getattr(exc, 'screen', None)
                stacktrace = getattr(exc, 'stacktrace', None)
            time.sleep(self._poll)
            if time.time() &gt; end_time:
                break
&gt;       raise TimeoutException(message, screen, stacktrace)
E       selenium.common.exceptions.TimeoutException: Message:

venv/lib64/python3.9/site-packages/selenium/webdriver/support/wait.py:80: TimeoutException

During handling of the above exception, another exception occurred:

self = &lt;test_validation_ui.TestUserInterfaceValidation object at 0x7f7f3da7d190&gt;
setup_ui_class = &lt;selenium.webdriver.chrome.webdriver.WebDriver (session="03c971acaa084890f609985623f58b0a")&gt;

--- a/ocs_ci/ocs/ui/base_ui.py
+++ b/ocs_ci/ocs/ui/base_ui.py
@@ -906,7 +906,7 @@ class SeleniumDriver(WebDriver):
     backoff=2,
     func=garbage_collector_webdriver,
 )
-def login_ui(console_url=None, username=None, password=None, **kwargs):
+def login_ui(console_url=None, username=None, password=None):
     """
     Login to OpenShift Console
 
@@ -984,13 +984,13 @@ def login_ui(console_url=None, username=None, password=None, **kwargs):
 
     if hci_platform_conf_confirmed:
         dashboard_url = console_url + "/dashboards"
-        # automatically proceed to load-cluster if test marked with provider decorator
-        if (
-            "request" in kwargs
-            and kwargs["request"].node.get_closest_marker("runs_on_provider")
-            and driver.current_url != dashboard_url
-        ):
+        # proceed to local-cluster page if not already there. The rule is always to start from the local-cluster page
+        # when the hci platform is confirmed and proceed to the client if needed from within the test
+        current_url = driver.current_url
+        logger.info(f"Current url: {current_url}")
+        if current_url != dashboard_url:
             # timeout is unusually high for different scenarios when default page is not loaded immediately
+            logger.info("Navigate to 'Local Cluster' page")
             navigate_to_local_cluster(
                 acm_page=locators[ocp_version]["acm_page"], timeout=180
             )
@@ -1070,11 +1070,13 @@ def navigate_to_local_cluster(**kwargs):
 
     all_clusters_dropdown = acm_page_loc["all-clusters_dropdown"]
     try:
+        logger.info("Navigate to Local Cluster page. Click all clusters dropdown")
         acm_dropdown = wait_for_element_to_be_visible(all_clusters_dropdown, timeout)
         acm_dropdown.click()
         local_cluster_item = wait_for_element_to_be_visible(
             acm_page_loc["local-cluster_dropdown_item"]
         )
+        logger.info("Navigate to Local Cluster page. Click local cluster item")
         local_cluster_item.click()
     except TimeoutException:
         wait_for_element_to_be_visible(acm_page_loc["local-cluster_dropdown"])

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -4917,6 +4917,19 @@ def setup_ui_class(request):
     return setup_ui_fixture(request)
 
 
+@pytest.fixture(scope="class")
+def setup_ui_class_factory(request):
+    # The problem with class scope fixtures is that they are executed always the first, when the class is loaded.
+    # This fixture is used to control fixture execution order, and call the fixture from within the test, after
+    # switch_to_provider fixture with autouse=True will be executed (for example).
+    # This way we can control the order of execution and perform login to management-console only after
+    # switching the context.
+    def factory():
+        setup_ui_fixture(request)
+
+    return factory
+
+
 @pytest.fixture(scope="function")
 def setup_ui(request):
     return setup_ui_fixture(request)
@@ -6698,9 +6711,11 @@ def cnv_dr_workload(request):
                     workload_pvc_selector=workload_details[
                         "dr_workload_app_pvc_selector"
                     ],
-                    appset_model=workload_details["appset_model"]
-                    if workload_type == constants.APPLICATION_SET
-                    else None,
+                    appset_model=(
+                        workload_details["appset_model"]
+                        if workload_type == constants.APPLICATION_SET
+                        else None
+                    ),
                 )
                 instances.append(workload)
                 total_pvc_count += workload_details["pvc_count"]

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -4686,7 +4686,7 @@ def login_factory_fixture(request):
     drivers = []
 
     def factory(username, password):
-        driver = login_ui(username=username, password=password, request=request)
+        driver = login_ui(username=username, password=password)
         drivers.append(driver)
         return driver
 
@@ -4936,7 +4936,7 @@ def setup_ui(request):
 
 
 def setup_ui_fixture(request):
-    driver = login_ui(request=request)
+    driver = login_ui()
 
     def finalizer():
         close_browser()

--- a/tests/cross_functional/ui/test_validation_ui.py
+++ b/tests/cross_functional/ui/test_validation_ui.py
@@ -39,7 +39,7 @@ class TestUserInterfaceValidation(object):
     @tier1
     @polarion_id("OCS-4925")
     @skipif_ui_not_support("validation")
-    def test_storage_cluster_validation_ui(self, setup_ui_class):
+    def test_storage_cluster_validation_ui(self, setup_ui_class_factory):
         """
         Validate Storage Cluster status on UI
 
@@ -47,6 +47,8 @@ class TestUserInterfaceValidation(object):
             setup_ui_class: login function on conftest file
 
         """
+        setup_ui_class_factory()
+
         validation_ui_obj = ValidationUI()
         validation_ui_obj.validate_storage_cluster_ui()
 
@@ -56,7 +58,7 @@ class TestUserInterfaceValidation(object):
     @bugzilla("2155743")
     @polarion_id("OCS-2575")
     @skipif_ui_not_support("validation")
-    def test_dashboard_validation_ui(self, setup_ui_class):
+    def test_dashboard_validation_ui(self, setup_ui_class_factory):
         """
         Validate User Interface of OCS/ODF dashboard
         Verify GET requests initiated by kube-probe on odf-console pod [cover bz-2155743]
@@ -66,6 +68,8 @@ class TestUserInterfaceValidation(object):
             setup_ui_class: login function on conftest file
 
         """
+        setup_ui_class_factory()
+
         ocs_version = version.get_semantic_ocs_version_from_config()
         if ocs_version >= version.VERSION_4_13:
             logger.info(
@@ -93,7 +97,7 @@ class TestUserInterfaceValidation(object):
     @polarion_id("OCS-4642")
     @skipif_ocs_version("<4.9")
     @skipif_ui_not_support("validation")
-    def test_odf_storagesystems_ui(self, setup_ui_class):
+    def test_odf_storagesystems_ui(self, setup_ui_class_factory):
         """
         Validate User Interface for ODF Storage Systems Tab for ODF 4.9
 
@@ -101,18 +105,21 @@ class TestUserInterfaceValidation(object):
             setup_ui_class: login function on conftest file
 
         """
+        setup_ui_class_factory()
+
         validation_ui_obj = ValidationUI()
         validation_ui_obj.odf_storagesystems_ui()
 
     @ui
     @tier1
+    @runs_on_provider
     @skipif_ocs_version("<4.9")
     @skipif_external_mode
     @skipif_mcg_only
     @pytest.mark.bugzilla("2096414")
     @polarion_id("OCS-4685")
     @skipif_hci_provider_or_client
-    def test_odf_cephblockpool_compression_status(self, setup_ui_class):
+    def test_odf_cephblockpool_compression_status(self, setup_ui_class_factory):
         """
         Validate Compression status for cephblockpool at StorageSystem details and ocs-storagecluster-cephblockpool
         are matching
@@ -121,6 +128,7 @@ class TestUserInterfaceValidation(object):
             setup_ui_class: login function on conftest file
 
         """
+        setup_ui_class_factory()
 
         storage_system_details = (
             ValidationUI()
@@ -147,10 +155,11 @@ class TestUserInterfaceValidation(object):
     @tier1
     @runs_on_provider
     @pytest.mark.bugzilla("1994584")
-    def test_ocs_operator_is_not_present(self, setup_ui_class):
+    def test_ocs_operator_is_not_present(self, setup_ui_class_factory):
         """
         Validate odf operator is present in the installed operator tab in ui.
         """
+        setup_ui_class_factory()
 
         validation_ui_obj = ValidationUI()
         (

--- a/tests/functional/object/mcg/ui/test_mcg_ui.py
+++ b/tests/functional/object/mcg/ui/test_mcg_ui.py
@@ -56,6 +56,7 @@ class TestStoreUserInterface(object):
 
     @ui
     @tier1
+    @runs_on_provider
     @skipif_disconnected_cluster
     @pytest.mark.parametrize(
         argnames=["kind", "provider", "region"],
@@ -71,7 +72,13 @@ class TestStoreUserInterface(object):
         ],
     )
     def test_store_creation_and_deletion(
-        self, setup_ui_class, cld_mgr, cloud_uls_factory, kind, provider, region
+        self,
+        setup_ui_class_factory,
+        cld_mgr,
+        cloud_uls_factory,
+        kind,
+        provider,
+        region,
     ):
         """
         Test creation and deletion of MCG stores via the UI
@@ -84,6 +91,8 @@ class TestStoreUserInterface(object):
         5. Verify store has been deleted via 'oc' cmd
 
         """
+        setup_ui_class_factory()
+
         log_step(
             "Navigate to Data Foundation / Object Storage / (Backing Store | Namespace Store)"
         )
@@ -290,6 +299,7 @@ class TestObcUserInterface(object):
 
     @ui
     @tier1
+    @runs_on_provider
     @bugzilla("2097772")
     @pytest.mark.parametrize(
         argnames=["storageclass", "bucketclass", "delete_via", "verify_ob_removal"],
@@ -324,7 +334,12 @@ class TestObcUserInterface(object):
         ],
     )
     def test_obc_creation_and_deletion(
-        self, setup_ui_class, storageclass, bucketclass, delete_via, verify_ob_removal
+        self,
+        setup_ui_class_factory,
+        storageclass,
+        bucketclass,
+        delete_via,
+        verify_ob_removal,
     ):
         """
         Test creation and deletion of an OBC via the UI
@@ -332,6 +347,8 @@ class TestObcUserInterface(object):
         The test covers BZ #2097772 Introduce tooltips for contextual information
         The test covers BZ #2175685 RGW OBC creation via the UI is blocked by "Address form errors to proceed"
         """
+        setup_ui_class_factory()
+
         obc_name = create_unique_resource_name(
             resource_description="ui", resource_type="obc"
         )

--- a/tests/functional/object/mcg/ui/test_namespace_store.py
+++ b/tests/functional/object/mcg/ui/test_namespace_store.py
@@ -31,9 +31,10 @@ class TestNamespaceStoreUI(object):
 
     @ui
     @tier1
+    @runs_on_provider
     @pytest.mark.bugzilla("2158922")
     @polarion_id("OCS-5125")
-    def test_create_namespace_store_ui(self, setup_ui_class, pvc_factory):
+    def test_create_namespace_store_ui(self, setup_ui_class_factory, pvc_factory):
         """
         1. Create a new PVC on openshift-storage namespce.
         2. Create namespacestore via ui based on filesystem and mount to new pvc
@@ -42,6 +43,8 @@ class TestNamespaceStoreUI(object):
         5. Delete PVC
 
         """
+        setup_ui_class_factory()
+
         self.namespace_store_obj = None
         openshift_storage_ns_obj = OCP(namespace=config.ENV_DATA["cluster_namespace"])
         pvc_obj = pvc_factory(

--- a/tests/functional/ui/test_pvc_ui.py
+++ b/tests/functional/ui/test_pvc_ui.py
@@ -73,7 +73,7 @@ class TestPvcUserInterface(object):
         self,
         project_factory,
         teardown_factory,
-        setup_ui_class,
+        setup_ui_class_factory,
         sc_name,
         access_mode,
         pvc_size,
@@ -83,6 +83,9 @@ class TestPvcUserInterface(object):
         Test create, resize and delete pvc via UI
 
         """
+
+        setup_ui_class_factory()
+
         # Creating a test project via CLI
         pro_obj = project_factory()
         project_name = pro_obj.namespace
@@ -239,7 +242,7 @@ class TestPvcUserInterface(object):
         self,
         project_factory,
         teardown_factory,
-        setup_ui_class,
+        setup_ui_class_factory,
         sc_name,
         access_mode,
         clone_access_mode,
@@ -248,6 +251,9 @@ class TestPvcUserInterface(object):
         Test to verify PVC clone from UI
 
         """
+
+        setup_ui_class_factory()
+
         pvc_size = "1"
         vol_mode = constants.VOLUME_MODE_FILESYSTEM
 

